open-apparel-registry/open-apparel-registry
# Support Front-end Environment Variables, Rollbar 2

## Overview

An alternative to #207, that instead of replacing Django-SPA with WhiteNoise, renames the generated entry file to `main.html` instead of `index.html`. This (rather brilliant) suggestion by @hectcastro allows us to bypass Django-SPA's default behavior and use `oar.urls` to render the page.

We still need to add a custom 404 handler, as by default Django-SPA tries to send non-matching routes to `static/index.html` which no longer exists.

This also preserves the behavior of `/admin/` which was broken in #207.

Supersedes #207 
Connects #64
Connects #203

## Motivation

The main reason to do this instead of having a settings endpoint that is queried after page load is that anything that requires an environment variable will have to check if the variable has loaded before it can function. As the number of environment variables grows, so will the unnecessary asynchronous code in the front-end. By ensuring that all variables are present at load, we simplify current and future code.

## Notes

The replacement of a 404 handler with a Django view potentially exposes us to DDoS attacks. However, if there are CDNs in front of the Django container, we should be okay.

~React errors are not caught because of [Error Boundaries in React 16](https://rollbar.com/blog/react-error-boundaries/). I'm going to make a follow up issue for fixing those.~ Fixed in a10ff7e.

## Testing Instructions

* Run the steps as in #207
* Ensure that [:8081/admin/](http://localhost:8081/admin/) loads and works correctly
